File: platforms/flutter/lib/src/display_list.dart
// display_list.dart — Dart mirror of ratex-types DisplayList / DisplayItem
//
// JSON uses serde internally-tagged format:
//   DisplayItem: {"type": "GlyphPath", "x": ..., ...}  (flat)
//   PathCommand: {"type": "MoveTo",    "x": ..., "y": ...}  (flat)

// MARK: - Top-level output

class DisplayList {
  final double width;
  final double height;
  final double depth;
  final List<DisplayItem> items;

  const DisplayList({
    required this.width,
    required this.height,
    required this.depth,
    required this.items,
  });

  factory DisplayList.fromJson(Map<String, dynamic> json) => DisplayList(
        width:  (json['width']  as num).toDouble(),
        height: (json['height'] as num).toDouble(),
        depth:  (json['depth']  as num).toDouble(),
        items:  (json['items'] as List<dynamic>)
            .map((e) => DisplayItem.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

// MARK: - Drawing commands (internally tagged: {"type": "GlyphPath", ...})

sealed class DisplayItem {
  const DisplayItem();

  factory DisplayItem.fromJson(Map<String, dynamic> json) {
    final type = json['type'] as String;
    switch (type) {
      case 'GlyphPath': return GlyphPathItem.fromJson(json);
      case 'Line':      return LineItem.fromJson(json);
      case 'Rect':      return RectItem.fromJson(json);
      case 'Path':      return PathItem.fromJson(json);
      default: throw FormatException('Unknown DisplayItem type: $type');
    }
  }
}

class GlyphPathItem extends DisplayItem {
  final double x, y, scale;
  final String font;
  final int charCode;
  final List<PathCommand> commands;
  final RaTeXColor color;

  const GlyphPathItem({
    required this.x, required this.y, required this.scale,
    required this.font, required this.charCode,
    required this.commands, required this.color,
  });

  factory GlyphPathItem.fromJson(Map<String, dynamic> j) => GlyphPathItem(
        x: (j['x'] as num).toDouble(), y: (j['y'] as num).toDouble(),
        scale: (j['scale'] as num).toDouble(),
        font: j['font'] as String,
        charCode: j['char_code'] as int,
        commands: (j['commands'] as List?)
                ?.map((e) => PathCommand.fromJson(e as Map<String, dynamic>))
                .toList() ??
            [],
        color: RaTeXColor.fromJson(j['color'] as Map<String, dynamic>),
      );
}

class LineItem extends DisplayItem {
  final double x, y, width, thickness;
  final RaTeXColor color;
  final bool dashed;

  const LineItem({required this.x, required this.y,
                  required this.width, required this.thickness,
                  required this.color, this.dashed = false});

  factory LineItem.fromJson(Map<String, dynamic> j) => LineItem(
        x: (j['x'] as num).toDouble(), y: (j['y'] as num).toDouble(),
        width: (j['width'] as num).toDouble(),
        thickness: (j['thickness'] as num).toDouble(),
        color: RaTeXColor.fromJson(j['color'] as Map<String, dynamic>),
        dashed: j['dashed'] as bool? ?? false,
      );
}

class RectItem extends DisplayItem {
  final double x, y, width, height;
  final RaTeXColor color;

  const RectItem({required this.x, required this.y,
                  required this.width, required this.height,
                  required this.color});

  factory RectItem.fromJson(Map<String, dynamic> j) => RectItem(
        x: (j['x'] as num).toDouble(), y: (j['y'] as num).toDouble(),
        width: (j['width'] as num).toDouble(),
        height: (j['height'] as num).toDouble(),
        color: RaTeXColor.fromJson(j['color'] as Map<String, dynamic>),
      );
}

class PathItem extends DisplayItem {
  final double x, y;
  final List<PathCommand> commands;
  final bool fill;
  final RaTeXColor color;

  const PathItem({required this.x, required this.y,
                  required this.commands, required this.fill,
                  required this.color});

  factory PathItem.fromJson(Map<String, dynamic> j) => PathItem(
        x: (j['x'] as num).toDouble(), y: (j['y'] as num).toDouble(),
        commands: (j['commands'] as List).map((e) => PathCommand.fromJson(e as Map<String, dynamic>)).toList(),
        fill: j['fill'] as bool,
        color: RaTeXColor.fromJson(j['color'] as Map<String, dynamic>),
      );
}

// MARK: - Path commands (internally tagged: {"type": "MoveTo", "x": ..., "y": ...})

sealed class PathCommand {
  const PathCommand();

  factory PathCommand.fromJson(Map<String, dynamic> json) {
    final type = json['type'] as String;
    switch (type) {
      case 'MoveTo':
        return MoveToCmd((json['x'] as num).toDouble(), (json['y'] as num).toDouble());
      case 'LineTo':
        return LineToCmd((json['x'] as num).toDouble(), (json['y'] as num).toDouble());
      case 'CubicTo':
        return CubicToCmd(
          (json['x1'] as num).toDouble(), (json['y1'] as num).toDouble(),
          (json['x2'] as num).toDouble(), (json['y2'] as num).toDouble(),
          (json['x']  as num).toDouble(), (json['y']  as num).toDouble(),
        );
      case 'QuadTo':
        return QuadToCmd(
          (json['x1'] as num).toDouble(), (json['y1'] as num).toDouble(),
          (json['x']  as num).toDouble(), (json['y']  as num).toDouble(),
        );
      case 'Close':
        return const CloseCmd();
      default:
        throw FormatException('Unknown PathCommand type: $type');
    }
  }
}

class MoveToCmd  extends PathCommand { final double x, y; const MoveToCmd(this.x, this.y); }
class LineToCmd  extends PathCommand { final double x, y; const LineToCmd(this.x, this.y); }
class CubicToCmd extends PathCommand {
  final double x1, y1, x2, y2, x, y;
  const CubicToCmd(this.x1, this.y1, this.x2, this.y2, this.x, this.y);
}
class QuadToCmd  extends PathCommand {
  final double x1, y1, x, y;
  const QuadToCmd(this.x1, this.y1, this.x, this.y);
}
class CloseCmd extends PathCommand { const CloseCmd(); }

// MARK: - Color

class RaTeXColor {
  final double r, g, b, a;
  const RaTeXColor(this.r, this.g, this.b, this.a);

  factory RaTeXColor.fromJson(Map<String, dynamic> j) => RaTeXColor(
        (j['r'] as num).toDouble(), (j['g'] as num).toDouble(),
        (j['b'] as num).toDouble(), (j['a'] as num).toDouble());

  /// Convert to a Flutter Color (32-bit ARGB int).
  int toFlutterColor() {
    final ai = (a * 255).round().clamp(0, 255);
    final ri = (r * 255).round().clamp(0, 255);
    final gi = (g * 255).round().clamp(0, 255);
    final bi = (b * 255).round().clamp(0, 255);
    return (ai << 24) | (ri << 16) | (gi << 8) | bi;
  }
}

File: platforms/ios/tests/Sources/RaTeXCore/DisplayList.swift
// DisplayList.swift — Swift mirror of ratex-types DisplayList / DisplayItem
//
// JSON uses serde internally-tagged format:
//   DisplayItem: {"type": "GlyphPath", "x": ..., "y": ..., ...}
//   PathCommand: {"type": "MoveTo",    "x": ..., "y": ...}

import Foundation

// MARK: - Top-level output

public struct DisplayList: Codable {
    /// DisplayList JSON protocol version (optional). Missing implies version 0.
    public let version: Int?
    /// Total width in em units.
    public let width: Double
    /// Ascent above baseline in em units.
    public let height: Double
    /// Descent below baseline in em units.
    public let depth: Double
    /// Ordered list of drawing commands.
    public let items: [DisplayItem]
}

// MARK: - Drawing commands

public enum DisplayItem: Codable {
    case glyphPath(GlyphPathData)
    case line(LineData)
    case rect(RectData)
    case path(PathData)
    /// Forward-compatibility: unknown item type; should be ignored by renderers/tests.
    case unknown(String)

    private enum TypeKey: String, CodingKey { case type }

    public init(from decoder: Decoder) throws {
        let tag = try decoder.container(keyedBy: TypeKey.self)
            .decode(String.self, forKey: .type)
        switch tag {
        case "GlyphPath": self = .glyphPath(try GlyphPathData(from: decoder))
        case "Line":      self = .line(try LineData(from: decoder))
        case "Rect":      self = .rect(try RectData(from: decoder))
        case "Path":      self = .path(try PathData(from: decoder))
        default:
            self = .unknown(tag)
        }
    }

    public func encode(to encoder: Encoder) throws {
        switch self {
        case .glyphPath(let d): try d.encode(to: encoder)
        case .line(let d):      try d.encode(to: encoder)
        case .rect(let d):      try d.encode(to: encoder)
        case .path(let d):      try d.encode(to: encoder)
        case .unknown(let tag):
            var c = encoder.container(keyedBy: TypeKey.self)
            try c.encode(tag, forKey: .type)
        }
    }
}

// MARK: - Item payloads (flat, include "type" key for round-trip)

public struct GlyphPathData: Codable {
    public let x: Double
    public let y: Double
    public let scale: Double
    public let font: String
    public let charCode: UInt32
    /// Placeholder paths; omitted in current serialized output. Default to empty when absent.
    public let commands: [PathCommand]
    public let color: RaTeXColor

    enum CodingKeys: String, CodingKey {
        case x, y, scale, font
        case charCode = "char_code"
        case commands, color
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        x = try c.decode(Double.self, forKey: .x)
        y = try c.decode(Double.self, forKey: .y)
        scale = try c.decode(Double.self, forKey: .scale)
        font = try c.decode(String.self, forKey: .font)
        charCode = try c.decode(UInt32.self, forKey: .charCode)
        commands = try c.decodeIfPresent([PathCommand].self, forKey: .commands) ?? []
        color = try c.decode(RaTeXColor.self, forKey: .color)
    }
}

public struct LineData: Codable {
    public let x: Double
    public let y: Double
    public let width: Double
    public let thickness: Double
    public let color: RaTeXColor
}

public struct RectData: Codable {
    public let x: Double
    public let y: Double
    public let width: Double
    public let height: Double
    public let color: RaTeXColor
}

public struct PathData: Codable {
    public let x: Double
    public let y: Double
    public let commands: [PathCommand]
    public let fill: Bool
    public let color: RaTeXColor
}

// MARK: - Path commands  (internally tagged: {"type": "MoveTo", "x": ..., "y": ...})

public enum PathCommand: Codable {
    case moveTo(x: Double, y: Double)
    case lineTo(x: Double, y: Double)
    case cubicTo(x1: Double, y1: Double, x2: Double, y2: Double, x: Double, y: Double)
    case quadTo(x1: Double, y1: Double, x: Double, y: Double)
    case close
    /// Forward-compatibility: unknown command type; should be ignored.
    case unknown(String)

    private enum TypeKey: String, CodingKey { case type }

    // Nested structs for decoding each variant's fields from the same flat container
    private struct XY:    Codable { let x, y: Double }
    private struct Cubic: Codable { let x1, y1, x2, y2, x, y: Double }
    private struct Quad:  Codable { let x1, y1, x, y: Double }

    public init(from decoder: Decoder) throws {
        let tag = try decoder.container(keyedBy: TypeKey.self)
            .decode(String.self, forKey: .type)
        switch tag {
        case "MoveTo":
            let d = try XY(from: decoder)
            self = .moveTo(x: d.x, y: d.y)
        case "LineTo":
            let d = try XY(from: decoder)
            self = .lineTo(x: d.x, y: d.y)
        case "CubicTo":
            let d = try Cubic(from: decoder)
            self = .cubicTo(x1: d.x1, y1: d.y1, x2: d.x2, y2: d.y2, x: d.x, y: d.y)
        case "QuadTo":
            let d = try Quad(from: decoder)
            self = .quadTo(x1: d.x1, y1: d.y1, x: d.x, y: d.y)
        case "Close":
            self = .close
        default:
            self = .unknown(tag)
        }
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: TypeKey.self)
        switch self {
        case .moveTo(let x, let y):
            try c.encode("MoveTo", forKey: .type)
            try XY(x: x, y: y).encode(to: encoder)
        case .lineTo(let x, let y):
            try c.encode("LineTo", forKey: .type)
            try XY(x: x, y: y).encode(to: encoder)
        case .cubicTo(let x1, let y1, let x2, let y2, let x, let y):
            try c.encode("CubicTo", forKey: .type)
            try Cubic(x1: x1, y1: y1, x2: x2, y2: y2, x: x, y: y).encode(to: encoder)
        case .quadTo(let x1, let y1, let x, let y):
            try c.encode("QuadTo", forKey: .type)
            try Quad(x1: x1, y1: y1, x: x, y: y).encode(to: encoder)
        case .close:
            try c.encode("Close", forKey: .type)
        case .unknown(let tag):
            try c.encode(tag, forKey: .type)
        }
    }
}

// MARK: - Color

/// RGBA color with components in [0, 1].
public struct RaTeXColor: Codable {
    public let r: Float
    public let g: Float
    public let b: Float
    public let a: Float

    public static let black = RaTeXColor(r: 0, g: 0, b: 0, a: 1)
}
